rclcpp/include/rclcpp/subscription.hpp

```diff
@@ -363,11 +363,31 @@ class Subscription : public SubscriptionBase
     void * loaned_message,
     const rclcpp::MessageInfo & message_info) override
   {
+    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
+      // In this case, the message will be delivered via intra process and
+      // we should ignore this copy of the message.
+      return;
+    }
+
```

Comment:
Can you rephrase the questions? I'm not sure I follow what you're asking.

Reply:
> Publisher cannot use LoanedMessage with intra_process_enable.

Yes.  
I only refer to code of handle_message() and not consider whether code is suitable for loan message.  
I will remove above codes.   

For intra-process case, waitable is used for trigger and `dispatch_intra_process()` will be called
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp#L151-L157

dispatch_intra_process() will call user's callback.   
So I think handle_message() and handle_loaned_message() cannot be called while intra-process is enabled.  

I don't know what case make handle_message() called while  intra-process communication.   
That is,  whether handle_message() needs to include below code.   
```c++
    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
      // In this case, the message will be delivered via intra process and
      // we should ignore this copy of the message.
      return;
    }
```

Reply:
I'm still not following exactly, I can see the parts you might be concerned with, but I don't follow the specific scenario, or at least I don't see the problem. Maybe we can discuss it in one of the working groups.

Reply:
I actually think the code removed in [565d683](https://github.com/ros2/rclcpp/pull/1927/commits/565d683ca4aa2ddac9f450a38b248939f673a6f7) was useful, but again that might be related to the misunderstanding I have with the above.

AFAICT, it is possible to get the data via loanded message and intra process, so the check to avoid double delivery is valid (I think).

Reply:
> I actually think the code removed in [565d683](https://github.com/ros2/rclcpp/pull/1927/commits/565d683ca4aa2ddac9f450a38b248939f673a6f7) was useful

agree, i believe this code should stay to avoid double delivery. this is what i am saying at the beginning though.

> I'm still not following exactly

sorry, let me add some more comments here. of course we can talk online.

even if publisher does not use `LoanedMessage` in ROS 2, underlying rmw implementation would use shared memory for inter-process communication. in this case, on subscription it could call `handle_loaned_message` if `can_loan_messages` is true. so w/o calling `matches_any_intra_process_publishers` in `handle_loaned_message`, there would be double delivery via rclcpp intra-process and `handle_loaned_message`.  i may be missing something, i am not sure if we can make this happen.

Reply:
I will check what case leads to double delivery.